cairo/src/contracts/token/extensions/hyp_erc20_vault_collateral.cairo

```diff
@@ -0,0 +1,242 @@
+
+    #[storage]
+    struct Storage {
+        vault: ERC4626ABIDispatcher,
+        #[substorage(v0)]
+        collateral: HypErc20CollateralComponent::Storage,
+        #[substorage(v0)]
+        mailbox: MailboxclientComponent::Storage,
+        #[substorage(v0)]
+        token_router: TokenRouterComponent::Storage,
+        #[substorage(v0)]
+        gas_router: GasRouterComponent::Storage,
+        #[substorage(v0)]
+        router: RouterComponent::Storage,
+        #[substorage(v0)]
+        ownable: OwnableComponent::Storage,
+        #[substorage(v0)]
+        upgradeable: UpgradeableComponent::Storage
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        #[flat]
+        HypErc20CollateralEvent: HypErc20CollateralComponent::Event,
+        #[flat]
+        MailBoxClientEvent: MailboxclientComponent::Event,
+        #[flat]
+        GasRouterEvent: GasRouterComponent::Event,
+        #[flat]
+        RouterEvent: RouterComponent::Event,
+        #[flat]
+        OwnableEvent: OwnableComponent::Event,
+        #[flat]
+        TokenRouterEvent: TokenRouterComponent::Event,
+        #[flat]
+        UpgradeableEvent: UpgradeableComponent::Event
+    }
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        mailbox: ContractAddress,
+        vault: ContractAddress,
+        owner: ContractAddress,
+        hook: Option<ContractAddress>,
+        interchain_security_module: Option<ContractAddress>
+    ) {
+        self.ownable.initializer(owner);
+        self.mailbox.initialize(mailbox, hook, interchain_security_module);
+        let vault_dispatcher = ERC4626ABIDispatcher { contract_address: vault };
+        let erc20 = vault_dispatcher.asset();
+        self.collateral.initialize(erc20);
+        self.vault.write(vault_dispatcher);
+    }
+
+    impl TokenRouterTransferRemoteHookImpl of TokenRouterTransferRemoteHookTrait<ContractState> {
+        fn _transfer_remote( // this overrides specific internal _transfer_remote check parameters ensure any additional thing is needed
+            ref self: TokenRouterComponent::ComponentState<ContractState>,
+            destination: u32,
+            recipient: u256,
+            amount_or_id: u256,
+            value: u256,
+            hook_metadata: Option<Bytes>,
+            hook: Option<ContractAddress>
+        ) -> u256 {
+            let mut contract_state = TokenRouterComponent::HasComponent::get_contract_mut(ref self);
+            TokenRouterHooksTraitImpl::transfer_from_sender_hook(ref self, amount_or_id);
+            let shares = contract_state._deposit_into_vault(amount_or_id);
+            let vault = contract_state.vault.read();
+            let exchange_rate = math::mul_div(
+                PRECISION, vault.total_assets(), vault.total_supply(),
+            );
+            let token_metadata: Bytes =
+                BytesTrait::new_empty(); // TODO: exchange_rate // abi.encode ? should it be big endian?
```

Comment:
@ametel01  what do you think about this? solidity does abi.encode(exchange_rate) [here](https://github.com/hyperlane-xyz/hyperlane-monorepo/blob/518a1bef91a025c5a2c580134aa5fdb48fdd5667/solidity/contracts/token/extensions/HypERC4626Collateral.sol#L55)